reactos/drivers/usb/usbhub_new/usbhub.c

```diff
@@ -0,0 +1,5047 @@
+                      &FailReason,
+                      sizeof(FailReason));
+
+        ZwClose(KeyHandle);
+    }
+
+    return Status;
+}
+
+VOID
+NTAPI
+USBH_UrbTimeoutDPC(IN PKDPC Dpc,
+                   IN PVOID DeferredContext,
+                   IN PVOID SystemArgument1,
+                   IN PVOID SystemArgument2)
+{
+    PUSBHUB_URB_TIMEOUT_CONTEXT HubTimeoutContext;
+    KIRQL OldIrql;
+    BOOL IsCompleted;
+
+    DPRINT("USBH_TimeoutDPC ... \n");
+
+    HubTimeoutContext = DeferredContext;
+
+    KeAcquireSpinLock(&HubTimeoutContext->UrbTimeoutSpinLock, &OldIrql);
+    IsCompleted = HubTimeoutContext->IsNormalCompleted;
+    KeReleaseSpinLock(&HubTimeoutContext->UrbTimeoutSpinLock, OldIrql);
+
+    if (!IsCompleted)
+    {
+        IoCancelIrp(HubTimeoutContext->Irp);
+    }
+
+    KeSetEvent(&HubTimeoutContext->UrbTimeoutEvent,
+               EVENT_INCREMENT,
+               FALSE);
+}
+
+NTSTATUS
+NTAPI
+USBH_SetPdoRegistryParameter(IN PDEVICE_OBJECT DeviceObject,
+                             IN PCWSTR ValueName,
+                             IN PVOID Data,
+                             IN ULONG DataSize,
+                             IN ULONG Type,
+                             IN ULONG DevInstKeyType)
+{
+    NTSTATUS Status;
+    UNICODE_STRING ValueNameString;
+    HANDLE KeyHandle;
+
+    DPRINT("USBH_SetPdoRegistryParameter ... \n");
+
+    RtlInitUnicodeString(&ValueNameString, ValueName);
+
+    Status = IoOpenDeviceRegistryKey(DeviceObject,
+                                     DevInstKeyType,
+                                     STANDARD_RIGHTS_ALL,
+                                     &KeyHandle);
+
+    if (NT_SUCCESS(Status))
+    {
+         ZwSetValueKey(KeyHandle,
+                       &ValueNameString,
+                       0,
+                       Type,
+                       Data,
+                       DataSize);
+
+        ZwClose(KeyHandle);
+    }
+
+    return Status;
+}
+
+NTSTATUS
+NTAPI
+USBH_SyncSubmitUrb(IN PDEVICE_OBJECT DeviceObject,
+                   IN PURB Urb)
+{
+    KEVENT Event;
+    IO_STATUS_BLOCK IoStatusBlock;
+    PIRP Irp;
+    PIO_STACK_LOCATION IoStack;
+    PUSBHUB_URB_TIMEOUT_CONTEXT HubTimeoutContext;
+    BOOLEAN IsWaitTimeout = FALSE;
+    LARGE_INTEGER DueTime;
+    NTSTATUS Status;
+
+    DPRINT("USBH_SyncSubmitUrb: ... \n");
+
+    Urb->UrbHeader.UsbdDeviceHandle = NULL;
+
+    KeInitializeEvent(&Event, NotificationEvent, FALSE);
+
+    Irp = IoBuildDeviceIoControlRequest(IOCTL_INTERNAL_USB_SUBMIT_URB,
+                                        DeviceObject,
+                                        NULL,
+                                        0,
+                                        NULL,
+                                        0,
+                                        TRUE,
+                                        &Event,
+                                        &IoStatusBlock);
+
+    if (!Irp)
+    {
+        Status = STATUS_INSUFFICIENT_RESOURCES;
```

Comment:
usbhub.c:273: You are missing a goto or return here, since the code below will dereference Irp